_drafts/swim.md

```diff
@@ -0,0 +1,441 @@
+distributed failure detector protocols are:
+
+* **Strong Completeness**: crash-failure of any group member is detected by all non-faulty members ;
+* **Speed of failure detection**: the time interval between a member failure and its detection by some non-faulty group member ;
+* **Accuracy**: the rate of false positives of failure detection (note: [100%
+  accuracy is impossible on asynchronous networks](http://www.ecommons.cornell.edu/bitstream/1813/7192/1/95-1535.pdf)) ;
+* **Network Message Load**, in bytes per second generated by the protocol.
+
+## Evaluating heartbeat
+
+In the light of our evaluation criteria, what can we say about heartbeating?
+
+* **Strong Completeness**: every nodes receives heartbeats, so a node failure
+    will be detected independently by every other node in the cluster ;
+* **Speed of failure detection**: this is tunable and is often set to a value
+    looking like **<code>T * limit</code>** which means that a failure will be
+    detected after <code>limit</code> intervals <code>T</code> of time ;
+* **Network Message Load**: **<code>O(n²)</code>**, where <code>n</code> is the
+    number of nodes in the cluster.
+* **Accuracy**: there is a trade-off to be made here. No network being perfect,
+    there will always be a portion of the packets that will be lost. So if the
+    <code>limit</code> value is set too low (to 1, for instance), a node might
+    be considered dead just because the heartbeat was dropped by the network.
+    On the other hand, if <code>limit</code> is set to a value that is too high,
+    the speed of failure detection will increase accordingly.
+    The **trade-off** is therefore **between speed and accuracy**. With a
+    reliable-enough network, we can safely assume that a low <code>limit</code>
+    will suffice and hence we consider heartbeating as having a "**reasonably
+    high**" accuracy.
+
+Back in 2002, a group of researchers from Cornell University (the same three as
+previously) devised a new protocol to address the network load imposed by
+traditional heart-beating protocols. They called their new protocol SWIM.
+
+## Enters *SWIM*
+
+The new idea introduced in *SWIM* — or **S**calable **W**eakly-consistent
+**I**nfection-style Process Group **M**embership Protocol — is the separation of
+the failure detection and membership update dissemination functionalities.
+
+### Failure detection component
+
+As the name suggests, the role of this component is to detect potential failures
+of nodes in the cluster.
+
+Like in heartbeating protocols, each node has a list of the nodes that are in
+the cluster. But instead of brutally pinging all of them every <code>T</code>
+interval of time, only one of them will be checked.
+
+On this schema, we see a node **randomly choosing** another one of its peers to
+check if it is still alive. It sends a <code>PING</code> message and expects an
+<code>ACK</code> in response, indicating that the target node is healthy.
```

Comment:
Is "UP" and "healthy" the same in this context? If a node ACK the message, is it enough to say he is healthy?

Reply:
Yep. As far as the protocol is concerned, replying = healthy. I guess that I could have talked about distinguishing nuances of health (like "I'm here, but consider me busy or overloaded for now) but the implications are complex, not handled by the protocol and the blogpost is probably already hard enough to read :D